pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1083,29 +1084,53 @@ func (n *NetworkPolicyController) syncAddressGroup(key string) error {
 	return nil
 }
 
-func (n *NetworkPolicyController) populateAddressGroupMemberSet(g *antreatypes.AddressGroup) controlplane.GroupMemberSet {
-	var pods []*v1.Pod
-	var externalEntities []*v1alpha2.ExternalEntity
+// getAddressGroupMemberSet knows how to construct a GroupMemberSet that contains
+// all the entities selected by an AddressGroup.
+func (n *NetworkPolicyController) getAddressGroupMemberSet(g *antreatypes.AddressGroup) controlplane.GroupMemberSet {
 	// Check if an internal Group object exists corresponding to this AddressGroup.
-	_, found, _ := n.internalGroupStore.Get(g.Name)
+	groupObj, found, _ := n.internalGroupStore.Get(g.Name)
 	if found {
 		// This AddressGroup is derived from a ClusterGroup.
-		pods, externalEntities = n.groupingInterface.GetEntities(clusterGroupType, g.Name)
-	} else {
-		pods, externalEntities = n.groupingInterface.GetEntities(addressGroupType, g.Name)
+		group := groupObj.(*antreatypes.Group)
+		return n.getClusterGroupMemberSet(group)
+	}
+	return n.getMemberSetForGroupType(addressGroupType, g.Name)
+}
+
+// getClusterGroupMemberSet knows how to construct a GroupMemberSet that contains
+// all the entities selected by a ClusterGroup. For ClusterGroup that has childGroups,
+// the members are computed as the union of all its childGroup's members.
```

Comment:
So we flatten nested groups in controller and send both the flattened parent groups and child groups (?) to agent?
This can be a simpler strategy, but meanwhile have we thought about the possibility of let agent flatten the groups?
@tnqn

Reply:
We don't send the parent Groups to the agent... whenever a parent Group is referenced in policies, the its corresponding appliedToGroup/addressGroup gets the flattened addresses, and it's the ATGrp/AddrGrp we send to the agent

Reply:
If letting agent flatten the groups, we need to have a controlplane clustergroup API to send their members and childgroups to agents. Maybe this could be discussed separately when we make group API generic for features to consume?

Reply:
Yes, we can think about CHILD group flattening in agent later.

Back to the current implementation, so when two NPs share the same ClusterGroup in the Peers, we create a single AppliedToGroup/AddressGroup? And we still use the normalized ID for the AppliedTo/Address groups?
This way might be ok for now, but when we support multiple groups in Peer or the new AppliedTo, it won't help to send a single copy of Group members to agent. We need to think about a solution for that.

Reply:
I think the problem of 1:1 mapping between AppliedTo/AddressGroup and CG is independent of nested groups: if we support multiple CGs in a single peer, even w/o nested group there will be this problem. And if that is the case, there's two level of flattening: 1. flattening nested group 2. flattening groups in a single peer for AppliedTo/AddressGroup. I still think 1 makes sense to be done in controller, even 2. The complexity of flattening can be taken care of by the controller and agent only needs to know about the result: it doesn't care about the nesting, does it?

Reply:
If we just have a single level of nesting, it is not complex to flatten, and in that case I would at least handle your #2 (flattening groups in a single peer for AppliedTo/AddressGroup) in agent, at least for AddressGroup (as AppliedTo group only includes local members so should be much smaller).
When a group is used by multiple policies/rules (that is the reason we introduce group right), sharing the group member message can reduce data on-wire a lot.

Reply:
Got it. Makes sense to me.